[.NET] Avoid heap alloc when using StringNames as key in a Dictionary

fLindahl opened the pull request:
`StringNames` in Godot C# are IEquatable, and overrides `GetHashCode()`, which calls `godot_string_name.movable`'s version of that method, which was not overridden. This led to heap allocations when e.g. calling the indexer in a Dictionary with `StringName` type as Key.

This PR just adds the (assuming) missing method, which seems to fix the issue.

fLindahl committed:
[.NET] Avoid heap allocation when using StringNames as key in a Collection.Dictionary.

Changed StringName GetHashCode to call godot_string_name.GetHashCode instead of godot_string_name's (which was not overridden) as this otherwise leads to heap allocations when e.g. calling the indexer in a Dictionary with `StringName` type as Key.

raulsntos reviewed (changes requested):
Thanks for contributing to the .NET module!

The `movable` types are only meant to be used to store them in classes (like `StringName`) where we can't store ref structs. None of them implement any API other than for conversion. So I think it was intentional to not implement `GetHashCode`.

However, as you noticed, we likely meant to call the `godot_string_name.GetHashCode` method (which is implemented) from the `StringName.GetHashCode` implementation. So the fix should be something like this:

```diff
diff --git a/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs b/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs
index 21d9ada127..a28230636f 100644
--- a/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs
+++ b/modules/mono/glue/GodotSharp/GodotSharp/Core/StringName.cs
@@ -161,7 +161,7 @@ namespace Godot
 
         public override int GetHashCode()
         {
-            return NativeValue.GetHashCode();
+            return NativeValue.DangerousSelfRef.GetHashCode();
         }
     }
 }

```

fLindahl commented:
> The `movable` types are only meant to be used to store them in classes (like `StringName`) where we can't store ref structs. None of them implement any API other than for conversion. So I think it was intentional to not implement `GetHashCode`.

That makes sense! I removed the GetHashCode method from movable and changed StringName to use the existing version in `godot_string_name` instead. Thanks!

raulsntos reviewed (approved):
Looks good to me, thanks!

clayjohn commented:
Looks great! The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉
